src/constraint_system/ecc/scalar_mul/variable_base/mod.rs

```diff
@@ -0,0 +1,164 @@
+pub fn variable_base_scalar_mul(
+    composer: &mut StandardComposer,
+    jubjub_var: Variable,
+    point: Point,
+) -> PointScalar {
+    // Turn scalar into bits
+    let raw_bls_scalar = *composer.variables.get(&jubjub_var).unwrap();
+    let scalar_bits_var = scalar_decomposition(composer, jubjub_var, raw_bls_scalar);
+
+    let identity = Point::identity(composer);
+    let mut result = identity;
+
+    for bit in scalar_bits_var.into_iter().rev() {
+        result = result.fast_add(composer, result);
+        let point_to_add = conditional_select_identity(composer, bit, point);
+        result = result.fast_add(composer, point_to_add);
+    }
+
+    PointScalar {
+        point: result,
+        scalar: jubjub_var,
+    }
+}
+
+/// If bit == 0, then return zero else return value
+/// This is the polynomial f(x) = x * a
+/// Where x is the bit
+fn conditional_select_zero(
+    composer: &mut StandardComposer,
+    bit: Variable,
+    value: Variable,
+) -> Variable {
+    // returns bit * value
+    composer.mul(Scalar::one(), bit, value, Scalar::zero(), Scalar::zero())
+}
+/// If bit == 0, then return 1 else return value
+/// This is the polynomial f(x) = 1 - x + xa
+/// Where x is the bit
+fn conditional_select_one(
+    composer: &mut StandardComposer,
+    bit: Variable,
+    value: Variable,
+) -> Variable {
+    let value_scalar = composer.variables.get(&value).unwrap();
+    let bit_scalar = composer.variables.get(&bit).unwrap();
+
+    let f_x_scalar = Scalar::one() - bit_scalar + (bit_scalar * value_scalar);
+    let f_x = composer.add_input(f_x_scalar);
+
+    composer.poly_gate(
+        bit,
+        value,
+        f_x,
+        Scalar::one(),
+        -Scalar::one(),
+        Scalar::zero(),
+        -Scalar::one(),
+        Scalar::one(),
+        Scalar::zero(),
+    );
+
+    f_x
+}
+
+// If bit == 0 choose identity, if bit == 1 choose point_b
+fn conditional_select_identity(
+    composer: &mut StandardComposer,
+    bit: Variable,
+    point_b: Point,
+) -> Point {
+    let x = conditional_select_zero(composer, bit, *point_b.x());
+    let y = conditional_select_one(composer, bit, *point_b.y());
+
+    Point { x, y }
+}
+
+fn scalar_decomposition(
+    composer: &mut StandardComposer,
+    witness_var: Variable,
+    witness_scalar: Scalar,
+) -> Vec<Variable> {
+    // Decompose the bits
+    let scalar_bits = scalar_to_bits(&witness_scalar);
+
+    // Add all the bits into the composer
+    let scalar_bits_var: Vec<Variable> = scalar_bits
+        .iter()
+        .map(|bit| composer.add_input(Scalar::from(*bit as u64)))
+        .collect();
+
+    // Take the first 252 bits
+    let scalar_bits_var = scalar_bits_var[0..252].to_vec();
+
+    // Now ensure that the bits correctly accumulate to the witness given
+    let mut accumulator_var = composer.zero_var;
+    let mut accumulator_scalar = Scalar::zero();
+
+    for (power, bit) in scalar_bits_var.iter().enumerate() {
```

Comment:
Consider using `fold` here if you're accumulating.

Reply:
Feel free to open an issue regarding optimising for loops with iterators.